MatrixSDKTests/MXCryptoTests.m

```diff
@@ -3422,6 +3428,65 @@ - (void)testIsRoomSharingHistory
     }];
 }
 
+#pragma mark Helpers
+
+/**
+ Manually restart the session and wait until a given room has finished syncing all state
+ 
+ Note: there is a lot of state update and sync going on when the session is started,
+ and integration tests often assume given state before it has finished updating. To solve
+ that this helper method makes the best guesses by observing global notifications
+ and adding small delays to ensure all updates have really completed.
+ */
+- (void)restartSession:(MXSession *)session
+    waitingForRoomId:(NSString *)roomId
+             success:(void (^)(MXRoom *))success
+             failure:(void (^)(NSError *))failure
+{
+    __block id observer;
+    
+    // First start the session
+    [session start:^{
+        
+        // Wait until we know that the room has actually been created
+        observer = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionNewRoomNotification
+                                                          object:nil
+                                                           queue:[NSOperationQueue mainQueue]
+                                                      usingBlock:^(NSNotification * notification) {
+            if ([notification.userInfo[kMXSessionNotificationRoomIdKey] isEqualToString:roomId])
+            {
+                [[NSNotificationCenter defaultCenter] removeObserver:observer];
+                
+                MXRoom *room = [session roomWithRoomId:roomId];
+                if (room)
+                {
+                    // Now wait until this room reports sync completion
+                    observer = [[NSNotificationCenter defaultCenter] addObserverForName:kMXRoomInitialSyncNotification
+                                                                                 object:nil
+                                                                                  queue:[NSOperationQueue mainQueue]
+                                                                             usingBlock:^(NSNotification * notification) {
+                        [[NSNotificationCenter defaultCenter] removeObserver:observer];
+                        
+                        // Even when sync completed, there are actually still a few async updates that happen (i.e. the notification
+                        // fires too early), so have to add some small arbitrary delay.
+                        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
```

Comment:
Kind of annoying to have to stall for a whole second, but can see why it is needed 🙁

Reply:
Yea it is indeed. I would fix the notification, but then have no test to proove that I did not break something else, so this should be passing first before any changes to production code.